### ISSUE TITLE
How to update a nested list/map structure

### ISSUE THREAD

**Issue by alexw668** [facebook/immutable-js#692]
Reading the API documents and a very limited sample of how to manipulate an immutable data structure, I cannot find an easy way to update a nested tree like structure as illustrated in the following:

```javascript
const nodes = List.of(
  Map({
    id: '10001',
    name: 'First 1',
    parent_id: '00001',
    nodes: List.of(
      Map({
        id: '20001',
        parent_id: '10001',
        name: 'First 1/Second 1',
        nodes: List.of()
      }),
      Map({
        id: '20002',
        parent_id: '10001',
        name: 'First 1/Second 2',
        nodes: List.of(
          Map({
            id: '30001',
            parent_id: '20002',
            name: 'First 1/Second 2/Third 1',
            nodes: List.of()              
          })
        )
      })
    )
  }),
  Map({
    id: '10002',
    name: 'First 2',
    parent_id: '00001',
    nodes: List.of(
      Map({
        id: '20003',
        parent_id: '10002',
        name: 'First 2/Second 1',
        nodes: List.of()
      })
    )
  })
);
```

This nested structure can go deep at any arbitrary level. `id` of each node is unique. I have a need to either insert a new or update an existing child node (whose `parent_id` matches its parent node's `id`) of an existing node.

For example, assuming I want to add a new node as a child node of the one named `First 1/Second 2` (id being `20002`)  with a `Map({id: '30002', name: 'First 1/Second 2/Third 2', parent_id: '20002'})`, how would I update the nodes? 

On the other hand, if I wanted to update an existing node (for example changing name of the node named `First 1/Second 2/Third 1` (id being `30001`), how would I do it? 

Basically the tasks at hand are: 1) find the parent node whose `id` matches parent_id of the given map data, 2) check the nodes contained in the parent node to decide either to add a new node or to update an existing node; 3) return the immutable nodes with the update applied. 

I would appreciate it if anyone can help.

Thanks.

**Comment by danielepolencic** [comment]
>  I want to add a new node as a child node of the one named First 1/Second 2 (id being 20002) with a Map({id: '30002', name: 'First 1/Second 2/Third 2', parent_id: '20002'})

```js
const newNode = Immutable.JS({id: '30002', name: 'First 1/Second 2/Third 2', parent_id: '20002'});
nodes.update(0, (node) => node.update('nodes', (nodes) => nodes.push(newNode)));
```

as for

> find the parent node whose id matches parent_id of the given map data

you need to walk down the DAG and find the element. Immutable.js doesn't provide any algorithm to do so. You can borrow some code online or create your own.

> check the nodes contained in the parent node to decide either to add a new node or to update an existing node

If you need to update a property, you can use the `merge` method instead of `update`.
